### DISCUSSION TITLE
Some ideas

### DISCUSSION THREAD

**Discussion by arysom** [selwynpolit/d9book#24] [Ideas]
- creating custom Entities , there's a lot in Drupal docs about it, but maybe you could share some of your experiences. There are certain situations, where someone asks me to create a webapp with relations, data types and all that and it occurs that I can't mimic standard mysql model, I have to be creative. There's timestamp field, but it doesnt fit and so on. And the relationship, I'm using entity reference field, but in the end I produce a relation that looks like it's reverted - not standard one to many, but rather many in one. Or maybe that's my point of view.

- creating apis with drupal, json vs rest, currently I'm into Next-Drupal - fine project, was afraid to  use jsonapi, but now a little less afraid. 

- performance stuff - lot of area to cover, from time to time I come back to this [book](https://www.oreilly.com/library/view/high-performance-drupal/9781449358013/), and it's still valid. Lot of ideas from it are precious.

**Comment by selwynpolit** [comment]
Thanks @arysom, I appreciate your thoughts.  Interestingly, I haven't found much use for creating custom entities.  Mostly I find that everything I've thought would be a good fit have actually done an amazing job just being unpublished nodes.  In the project I am currently working on, we discussed it again.  The architect was leaning towards using a custom table but in the end, we all felt like having the data in nodes had too many benefits e.g. using views, direct editing etc.

I'll ponder more on the other issues.  I have more info on caching for sure and did play a fair amount with Varnish and Fastly.  Not sure my notes are that useful tho.

**Reply by arysom** [reply]
The entites seem a way to go when there's a lot of data to process, they always promise, that this is less resources consuming. But in the end you always add handlers, views handler, listbuilder, define fields in code and so on. You can leave bundles behind.

Once I saw a great combination in a system that used Commerce module. There was a lot of geographical data and there were classes - even not Entity, I think it was just for OOP comfort, but... there was a separate storage class which created the Entity from a node and set the data to cache. I still don't quite understand what was happening there.

Thank you for adding me in the credits, I'll try to add something more if I find I can. I need to read the whole book. It will take some time.
